Auth0/CredentialsManager.swift

```diff
@@ -399,49 +585,111 @@ public struct CredentialsManager {
                                 callback(.failure(CredentialsManagerError(code: .storeFailed)))
                             } else {
-                                self.dispatchGroup.leave()
+                                dispatchGroup.leave()
                                 callback(.success(newCredentials))
                             }
                         case .failure(let error):
-                            self.dispatchGroup.leave()
+                            dispatchGroup.leave()
                             callback(.failure(CredentialsManagerError(code: .renewFailed, cause: error)))
                         }
                     }
             }
 
-            self.dispatchGroup.wait()
+            dispatchGroup.wait()
+        }
+    }
+
+    // swiftlint:disable:next function_body_length function_parameter_count
+    private func retrieveAPICredentials(audience: String,
+                                        scope: String?,
+                                        minTTL: Int,
+                                        parameters: [String: Any],
+                                        headers: [String: String],
+                                        callback: @escaping (CredentialsManagerResult<APICredentials>) -> Void) {
+        let dispatchGroup = DispatchGroup()
+
+        self.dispatchQueue.async {
+            dispatchGroup.enter()
+
+            DispatchQueue.global(qos: .userInitiated).async {
+                if let apiCredentials = self.retrieveAPICredentials(audience: audience),
+                      !self.hasExpired(apiCredentials.expiresIn),
+                      !self.willExpire(apiCredentials.expiresIn, within: minTTL),
+                      !self.hasScopeChanged(from: apiCredentials.scope, to: scope) {
+                    dispatchGroup.leave()
+                    return callback(.success(apiCredentials))
+                }
+                guard let currentCredentials = self.retrieveCredentials() else {
+                    dispatchGroup.leave()
+                    return callback(.failure(.noCredentials))
+                }
+                guard let refreshToken = currentCredentials.refreshToken else {
+                    dispatchGroup.leave()
+                    return callback(.failure(.noRefreshToken))
+                }
+
+                self.authentication
+                    .renew(withRefreshToken: refreshToken, audience: audience, scope: scope)
+                    .parameters(parameters)
+                    .headers(headers)
+                    .start { result in
+                        switch result {
+                        case .success(let credentials):
+                            let newCredentials = Credentials(from: currentCredentials,
+                                                             idToken: credentials.idToken,
+                                                             refreshToken: credentials.refreshToken ?? refreshToken)
```

Comment:
The ID token and the refresh token (if present) are replaced in the app credentials, which are then saved with these updated values.